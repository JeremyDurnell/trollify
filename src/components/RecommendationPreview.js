import React, { Component } from "react";
import styled from "react-emotion";

import Recommendation from "./Recommendation";

const PreviewContainer = styled("div")`
  height: 60vh;
  width: 75%;
  margin: 5vh 5%;
  background-color: black;
  overflow-y: scroll;
`;

export default class RecommendationPreview extends Component {
  constructor(props) {
    super(props);

    this.state = { activeMusicIndex: 0, play: false };

    this.handleSelect = this.handleSelect.bind(this);
    this._playMusic = this._playMusic.bind(this);
    this.handleToggle = this.handleToggle.bind(this);
  }

  handleSelect(index, event) {
    event.stopPropagation();

    this.state.activeMusicIndex === index
      ? this.handleToggle()
      : this._playMusic(index);
  }

  _playMusic(index) {
    this.setState(
      {
        activeMusicIndex: index,
        play: true
      },
      () => {
        this.audioContainer.currentTime = 0;
        this.audioContainer.play();
      }
    );
  }

  handleToggle() {
    this.state.play ? this.audioContainer.pause() : this.audioContainer.play();
    this.setState({ play: !this.state.play });
  }
  render() {
    console.log(this.props);
    const recommendations =
      this.props.recommendations.length > 1 &&
      this.props.recommendations.map((recommendation, index) => (
        <Recommendation
          recommendation={recommendation}
          key={recommendation.name}
          index={index}
          handleSelect={this.handleSelect}
          playing={this.state.activeMusicIndex === index && this.state.play}
        />
      ));
    const activeMusic =
      this.props.recommendations.length > 1 &&
      this.props.recommendations[this.state.activeMusicIndex];
    return (
      <PreviewContainer>
        <audio
          autoPlay={this.state.play}
          preload="none"
          ref={ref => {
            this.audioContainer = ref;
          }}
          src={activeMusic.preview_url}
        />
        {recommendations}
      </PreviewContainer>
    );
  }
}
