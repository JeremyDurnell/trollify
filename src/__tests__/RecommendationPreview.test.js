import React from "react";
import ReactDOM from "react-dom";
import RecommendationPreview from "../components/RecommendationPreview";
import renderer from "react-test-renderer";

const recommendations = [{ name: "test" }];

it("renders without crashing", () => {
  const div = document.createElement("div");
  ReactDOM.render(
    <RecommendationPreview recommendations={recommendations} />,
    div
  );
  ReactDOM.unmountComponentAtNode(div);
});

it("renders correctly", () => {
  const tree = renderer
    .create(<RecommendationPreview recommendations={recommendations} />)
    .toJSON();
  expect(tree).toMatchSnapshot();
});

const testApp = renderer
  .create(<RecommendationPreview recommendations={recommendations} />)
  .getInstance();

it("has correct initial state", () => {
  expect(testApp.state.play).toBe(false);
  expect(testApp.state.activeMusicIndex).toBe(0);
});

it("has correct proptype", () => {
  expect(testApp.props.recommendations).toEqual(
    expect.arrayContaining([{ name: "test" }])
  );
});

it("toggles play in state correctly", () => {
  const audioTagMock = {
    play: function() {
      return null;
    },
    pause: function() {
      return null;
    }
  };

  testApp.audioContainer = audioTagMock;
  testApp.handleToggle();
  expect(testApp.state.play).toBe(true);

  testApp.audioContainer = audioTagMock;
  testApp.handleToggle();
  expect(testApp.state.play).toBe(false);
});
