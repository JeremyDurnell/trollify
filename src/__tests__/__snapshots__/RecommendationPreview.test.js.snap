// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<div
  className="css-1jreqdg"
>
  <audio
    preload="none"
    src={undefined}
  />
</div>
`;
